File: 29-04-2024/Remove every kth node.py
#Your task is to complete this function
#Your function should return the new head pointer
'''
class node:
    def __init__(self,x):
        self.data = x
        self.next = None
'''

class Solution:
    def deleteK(self, head, k):
      #Your Code Start
        if k == 1:  # If k is 1, every node needs to be deleted, so return None
            return None

        dummy = node(0)  # Create a dummy node to handle edge case when head needs to be deleted
        dummy.next = head
        prev = dummy
        current = head
        count = 0

        while current:
            count += 1

            if count % k == 0:  # If count is a multiple of k, delete the current node
                prev.next = current.next
            else:
                prev = current

            current = current.next

        return dummy.next
        #Your Code End


#{ 
 # Driver Code Starts
class node:

    def __init__(self, x):
        self.data = x
        self.next = None


def createLinkedList(arr):
    head = node(arr[0])
    curr = head
    for i in range(1, len(arr)):
        new_node = node(arr[i])
        curr.next = new_node
        curr = curr.next

    return head


def printlist(ptr):
    while ptr != None:
        print(ptr.data, end=" ")
        ptr = ptr.next
    print()


if __name__ == '__main__':
    t = int(input())
    for i in range(t):
        n = int(input())
        arr = list(map(int, input().strip().split()))
        k = int(input())

        obj = Solution()
        head = createLinkedList(arr)
        new_head = obj.deleteK(head, k)
        printlist(new_head)

# } Driver Code Ends

File: 29-04-2024/Remove every kth node.cpp
//{ Driver Code Starts
// C program to find n'th Node in linked list
#include <stdio.h>
#include <stdlib.h>
#include<iostream>
using namespace std;
/* Link list Node */
struct Node
{
    int data;
    struct Node* next;
    
    Node(int x){
        data = x;
        next = NULL;
    }
    
};


void append(struct Node** head_ref, struct Node **tail_ref, int new_data)
{
    struct Node* new_node = new Node(new_data);
    
    if (*head_ref == NULL)
       *head_ref = new_node;
    else
       (*tail_ref)->next = new_node;
    
    *tail_ref = new_node;
}


// } Driver Code Ends
/* Link list Node 

struct Node
{
    int data;
    struct Node* next;
    
    Node(int x){
        data = x;
        next = NULL;
    }
    
};
*/

/*You are required to complete this method*/

/* Function to get the middle of the linked list*/
/*K will always be in range */

class Solution {
    public:
    Node* deleteK(Node *head,int K){
      //Your code Start
       if (K == 1) { // If K is 1, simply return NULL as every node needs to be deleted
            return NULL;
        }

        Node *current = head;
        Node *prev = NULL;
        int count = 0;

        while (current != NULL) {
            count++;

            if (count % K == 0) { // If count is a multiple of K, delete the current node
                if (prev != NULL) {
                    prev->next = current->next;
                    delete current;
                    current = prev->next;
                } else { // If prev is NULL, it means head needs to be updated
                    head = current->next;
                    delete current;
                    current = head;
                }
            } else {
                prev = current;
                current = current->next;
            }
        }

        return head;
      //Your Code End
    }
};


//{ Driver Code Starts.

/* Driver program to test above function*/
int main()
{
  int T,i,n,l;
    cin>>T;
    while(T--){
    struct Node *head = NULL,  *tail = NULL;
        cin>>n;
        for(i=1;i<=n;i++)
        {
            cin>>l;
            append(&head, &tail, l);
        }
        int K;
        cin>>K;
        Solution obj;
        Node *res = obj.deleteK(head,K);
        Node *temp = res;
        while(temp!=NULL)
        {
            cout<<temp->data<<" ";
            temp=temp->next;
        }
        cout<<endl;
    }
    return 0;
}

// } Driver Code Ends
